MADE/made.jl

```diff
@@ -0,0 +1,140 @@
+#=
+Implements Masked AutoEncoder for Density Estimation, by Germain et al. 2015
+Re-implementation by Andrej Karpathy based on https://arxiv.org/abs/1502.03509
+Re-Re-implementation by Tejan Karmali using Flux.jl ;)
+=#
+
+using Flux
+using Flux: glorot_uniform, @epochs
+using NPZ
+
+# ------------------------------------------------------------------------------
+
+add_dims_r(a) = reshape(a, size(a)..., 1)
+add_dims_l(a) = reshape(a, 1, size(a)...)
+
+# ------------------------------------------------------------------------------
+
+struct MaskedDense{F,S,T,M}
+  # same as Linear except has a configurable mask on the weights
+  W::S
+  b::T
+  mask::M
+  σ::F
+end
+
+function MaskedDense(in::Integer, out::Integer; σ = identity)
+  return MaskedDense(param(glorot_uniform(out, in)), param(zeros(out)), ones(out, in), σ)
+end
+
+Flux.treelike(MaskedDense)
+
+function (a::MaskedDense)(x)
+  a.σ.(a.mask .* a.W * x .+ a.b)
+end
+
+function set_mask(a::MaskedDense, mask)
+  a.mask = mask
+end
+
+# ------------------------------------------------------------------------------
+
+mutable struct MADE
+  nin::Integer
+  nout::Integer
+  hidden_sizes::Array{Integer, 1}
+  net::Chain
+
+  # seeds for orders/connectivities of the model ensemble
+  natural_ordering::Bool
+  num_masks
+  seed::UInt  # for cycling through num_masks orderings
+
+  m::Dict
+
+  function MADE(in::Integer, hs, out::Integer, nat_ord::Bool, num_masks = 1)
+    # define a simple MLP neural net
+    hs = push!([in], hs...)
+    layers = [MaskedDense(hs[i], hs[i + 1]; σ = relu) for i = 1:length(hs) - 1]
+
+    net = Chain(layers..., MaskedDense(hs[end], out))
```

Comment:
Couldn't this have a special case for the first and last layers (rather than just the last one), allowing you to avoid the `push!` in line 57?

Reply:
Wouldn't it be an issue if `hs = []`? In that case, `Chain(MaskedDense(in, hs[1]), layers..., MaskedDense(hs[end], out))` gives `BoundsError`